### DISCUSSION TITLE
Data value JSON format

### DISCUSSION THREAD

**Discussion by larshelge** [dhis2/notes#324] [Ideas]
## Context

From a client and in particular the new data entry web app perspective, the current data value format is cumbersome to use, as it requires the client to specify the category option combo (and until recently, the attribute option combo) specifically. The option combos could be considered internal details which should not be exposed to the client. Having to load all option combos to the client uses unnecessary time and bandwidth.

## Considerations

JSON is not suited for custom property names in the same way XML is, meaning we need a fixed format for the category combination and category options. The naming of "combo" and "attribute" is not ideal, but we are constrained by the existing model and should not make drastic changes in format.

## Format

Proposed data value JSON format can be seen below. The new properties will be supported both for the individual and bulk format. The main goal is to avoid having to expose category option combinations, and instead rely on category combo and category options.

```json
{
  "dataElement": "fbfJHSPpUQD",
  "period": "202201",
  "orgUnit": "DiszpKrYNg8",
  "category": {
    "combo": "jCNGsC2NawV",
    "options": [
      "apsOixVZlf1",
      "zs2Ra6sikup"
    ]
  },
  "attribute": {
    "combo": "O4VaNks6tta",
    "options": [
      "CW81uF03hvV",
      "i4Nbp8S2G6A"
    ]
  },
  "value": "10"
}
```

**Comment by vietnguyen** [comment]
Link to doc of [current format](https://docs.dhis2.org/fr/develop/using-the-api/dhis-core-version-master/data.html)

**Comment by vietnguyen** [comment]
Changing the class `org.hisp.dhis.dxf2.datavalue.DataValue` will affect both `api/dataValues` and `api/dataValueSets` endpoints. 

But in `api/dataValueSet` endpoint  should we  also have an outer `attribute` object ?
```json
{
  "attribute": {
    "combo": "O4VaNks6tta",
    "options": [
      "CW81uF03hvV",
      "i4Nbp8S2G6A"
    ]
  },
  "dataValues": [
    {
      "dataElement": "fbfJHSPpUQD",
      "period": "202201",
      "orgUnit": "DiszpKrYNg8",
      "value": "10",
      "category": {
        "combo": "jCNGsC2NawV",
        "options": [
          "apsOixVZlf1",
          "zs2Ra6sikup"
        ]
      }
    },
    {
      "dataElement": "fbfJHSPpUQD",
      "period": "202201",
      "orgUnit": "DiszpKrYNg8",
      "value": "11",
      "category": {
        "combo": "jCNGsC2NawV",
        "options": [
          "apsOixVZlf1",
          "zs2Ra6sikup"
        ]
      }
    }
  ]
}
```

**Reply by larshelge** [reply]
Good point. I think we need to have that. Though I think we could potentially split this into two iterations and only support it for the inner data value object (and the single data value endpoint) in first iteration.

**Comment by bobjolliffe** [comment]
Lars I wonder why you want to expose the "combo" uid instead of the individual categories.  In a sense that is an internal detail like the categoryoptioncombo.  Which seems to require the order of the options which follow to be significant.  And I dont think the order of categories within a categorycombo is really significant.  It might be I am misunderstanding.  But wouldn't something similar to the following be more robust (and also easily parseable):`

```json
"dataValues": [
  {
    "dataElement": "fbfJHSPpUQD",
    "period": "202201",
    "orgUnit": "DiszpKrYNg8",
    "value": "10",
    "categories": [
      { "category": "jCNGsC2NawV", "option": "apsOixVZlf1"},
      { "category": "huhyy8yuhhl", "option": "zs2Ra6sikup"}
    ]
  }
]
```


Note that this would also be directly mappable" to  ADX. which I know is not a design goal here but might have some advantages.  For exdample, using codes instead of uids would render something like:       

```json
{
  "dataElement": "MALARIA_CASES",
  "period": "202201",
  "orgUnit": "DiszpKrYNg8",
  "value": "10",
  "categories": [
    { "category": "SEX", "option": "FEMALE"},
    { "category": "AGE", "option": "P0Y--P1Y"}
  ]
}
```

**Reply by Mohammer5** [reply]
I was wondering about this as well. When looking at the category/attribute option combos, the connection between categories and their option is the ordering in their array as categories have their own array and so do the options.
Mapping them directly removes confusion as to whether there is a connection between the categories and the options or not or if the order doesn't matter and the backend automatically knows which option belongs to which category (which is not the case!)

**Reply by bobjolliffe** [reply]
I didn't do attributes but they would basically be the same.  An array of "category,option".  Which could be on inner datavalue or outer datavalueset (or a mixture)

**Reply by jbee** [reply]
I like bob's suggestion. This makes most sense from a users point of view and I think from what I understand that he is right about the combo not having any order, it should be a unique set of categories which we can find using the categories.

**Reply by jbee** [reply]
One note on the format itself.
```json
"categories": [
      { "category": "jCNGsC2NawV", "option": "apsOixVZlf1"},
      { "category": "huhyy8yuhhl", "option": "zs2Ra6sikup"}
    ]
```
Could also be represented as:
```json
"categories": {
  "jCNGsC2NawV": "apsOixVZlf1",
  "huhyy8yuhhl": "zs2Ra6sikup"
}
```
Just to save us from the repetition of fields. Might not be important for 1 value but if you have many like this.

**Reply by bobjolliffe** [reply]
Yes that would work.  I was mindful of Lars's condition that "JSON is not suited for custom property names in the same way XML is".  But otherwise the compressed version looks fine to me.  Note the number of categories is probably typically between 0 and 3.

**Reply by Mohammer5** [reply]
```
"categories": {
  "jCNGsC2NawV": "apsOixVZlf1",
  "huhyy8yuhhl": "zs2Ra6sikup"
}
```

This shape would be ideal as that's how we store the selection in the new data entry app

**Reply by larshelge** [reply]
Thanks @bobjolliffe yes I was thinking about that too. The challenge is that in our model, i) having multiple category combos with the same categories is allowed and possible to create in the maintenance app, and ii) we do not re-use category option combinations across category combos.

Both of these are likely design mistakes, but nonetheless the way it works now.

We could add new constraints to the system starting now, but there is still a good chance that existing implementations/databases could be problematic.

So I went with `combo` since our model does not guarantee that a set of categories matches a single category combo.

Of course, we could say that in order to use this format, your categories must refer to a single category combo, if not we just return a warning in the API, similar to if your category option ID does not refer to an existing category option. I agree that the format is cleaner and exposes less internal detail, though we are in fact then returning warnings for a configuration which is perfectly valid in the current model.

**Reply by vietnguyen** [reply]
I assume we are also going to change the format of data value export to be the same ?

**Reply by Birkbjo** [reply]
I agree that this makes sense from a consumer point of view, but unfortunately I don't think it's explicit enough. You are currently able to create `categoryCombos` with the same categories multiple times. This way you would not know what catCombo you actually want to refer to.

EDIT: I see that Lars was way ahead of me here...

**Reply by jbee** [reply]
Is there any real use case to have same set of categories defined multiple times? Are people using this as a work-around or is this just unintentional duplication?

**Reply by larshelge** [reply]
No, not a useful use-case from what I know.

**Reply by abyot** [reply]
Yes, we could end-up with multiple category combos for the same set of categories. However, since we have a data element which is linked only to a single category combo, I think we will be fine with just categories and corresponding options.

**Reply by jbee** [reply]
Maybe we can also work towards adding a unique constraint or error messages to make this into an actual system property?

**Reply by bobjolliffe** [reply]
Lars it strikes me that the possibility of having multiple identical categorycombos is an unfortunate one, and rather than have it dictate the design we might hope to eradicate it instead.  I wonder how often this actually occurs in practice and could it be detected and rectified as part of metadata assessment and cleanup scripts?  I think there is also a different logical problem with the original proposal.  maybe also an edge case of sorts.  It is quite possible for the same category options to appear in the different categories in a categorycombo.  I think particularly of generic labels like YES, NO, UNKNOWN, UNSURE, UNTESTED, etc etc but your mind could wander and think of more.  So it is quite possible to have two categoryoptions and a categorycombo, but not be able to say which category each option relates to.  The only way to be absolutely unambiguous would be to combine both approaches and specify both the categorycombo and the category/option tuples.  Which starts to get verbose.  Probably it would be better to try and (i) eliminate the multiple identical categories or (ii) rely on what Abyot has pointed out, that the categorycombo is linked to the dataelement (though I think it can be overridden in the dataset so need to be careful).

**Comment by Birkbjo** [comment]
Note that my remarks below are in the context of the Data Entry app. 

Thanks Lars! It would be great to not need to "match" `options` to their `categoryOptionCombo` on the client. However, we would still need a way to uniquely an input field. In our current implementation we closely mirror how this was done in the old Data Entry app; by concatenating the `dataElementId` and `categoryOptionComboId`. This is also how it's done in the custom-forms. So this API-change would also need an update to how the custom-forms are handled, right? 

My first idea is to just concatenate the optionIds - eg. `dataElementId-optId1.optId2`. 

This do however make the order of the optionIds significant. This is not necessarily a problem on our client-side generated form, as that is mostly internal to our rendering logic, and we can work around that. But we would still prefer the options returned in this API to be stable.

**Reply by larshelge** [reply]
Right. We could continue to support `categoryOptionCombo` as a field in the new format too, so that the client could decide. Would that work, so that we can continue to handle custom forms the way we do now?
